### ISSUE TITLE
Sprout supports fewer subscribers than the target

### ISSUE THREAD

**Issue by rkd-msw** [Metaswitch/sprout#1674]
#### Symptoms

The load profile I'm trying to get is 120,000 subscribers each making 1.3 calls/hr and sending 2 REGISTERs/hr, but in order to avoid needing to spend the time registering that many subscribers. I'm using 1% of the number of subscribers doing 100x. I'd expect to get 144,144 subscribers taking 60% CPU on one node.

However, with 12,000 subscribers each making 20 REGISTERs/hr and 13 calls/hr, my Sprout node uses 65.55% CPU (34.45% idle), suggesting 113,000 subscribers per Sprout node. This is split 32.83% usr, 18.48% system and 14.03% softirq handling.

I've got pidstat output showing Sprout taking 104.75% CPU (out of a total 200%), Chronos 6.71%, rsyslogd 1.31%, memcached 1.03% and snmpd 1.03%.

http://abra:8080/job/stress/84/ has all the output files.

#### Impact

Calls will fail as Sprout gets overloaded before hitting its expected capacity.

#### Release and environment 

10.0.0-161216-~255.00.0-161216.020317

#### Steps to reproduce

Set up a deployment and run stress through it (using https://github.com/Metaswitch/clearwater-perf-utils), monitoring CPU usage with tools like sar.

**Comment by chris-elford-metaswitch** [comment]
I found the block text above a little hard to parse. I think the situation is as follows:

Targets
--------
| Subs | Calls/hr | Reg/hr | Target CPU% |
| ---- | -------- | ------ | ----------- |
| 120,000 | 1.3 | 2 | |
| 144,144 | | | 60 |
| 12,000 | 13 | 20 | |

Observed
----------

| Subs | Calls/hr | Reg/hr | Observed CPU% |
| ---- | -------- | ------ | ------------- |
| 12,000 | 13 | 20 | 65.55 |

**Comment by rkd-msw** [comment]
Not quite:

Targets
--------
| Subs | Calls/hr | Reg/hr | Target CPU% |
| ---- | -------- | ------ | ----------- |
| 144,144 | 1.3 | 2 | 60 |


Observed on 10.0.0-161216-~255.00.0-161216.020317
--------------------

| Subs | Calls/hr | Reg/hr | Observed CPU% |
| ---- | -------- | ------ | ------------- |
| 12,000 | 13 | 20 | 65.55 |

Extrapolated from 10.0.0-161216-~255.00.0-161216.020317 numbers
--------------------

| Subs | Calls/hr | Reg/hr | Extrapolated CPU% |
| ---- | -------- | ------ | ------------- |
| 144,144 | 1.3 | 2 | 78.73866 |

**Comment by chris-elford-metaswitch** [comment]
I think you've worked out the extrapolated figure by:

`(65.55 / 12000) * 144144 = 78.74`

(Let's not introduce spurious accuracy.)

An error of 1.6% would tip us over 80% CPU usage. That's not a big error margin, so I'm uncomfortable assuming that our CPU load scales linearly here. Can we confirm that 144,14 subs at 13/20 keeps CPU under 80% with another load test?

**Comment by rkd-msw** [comment]
I'm currently seeing only 55% CPU with 144,144 subs, which is new and surprising. I'll dig into this a little more, including retesting on 10.0.0-161216-~255.00.0-161216.020317 to see if I can repro the earlier numbers.

**Comment by rkd-msw** [comment]
I'm seeing about 55% idle CPU (not used CPU - 45% used) on 10.0.0-170112-~255.00.0-170112.020313 and  10.0.0-170218-~255.00.0-170218.020404. See http://abra:8080/job/stress/223/ and http://abra:8080/job/stress/222.

https://github.com/Metaswitch/sprout/compare/release-113...release-114 suggests that the most interesting change between 161216 and 170112 is https://github.com/Metaswitch/clearwater-net-snmp/pull/12 - we think that that caused a tight loop in Chronos so I can believe it caused high CPU usage in Sprout too.

If I can repro high CPU usage on the 161216 build when I test tomorrow, I'll close this. (I might close it anyway - I'm consistently getting good CPU% numbers now and I can't see what other investigation to do here.)

**Comment by rkd-msw** [comment]
I can't repro this on any build (e.g. http://abra:8080/job/stress-fast/2/), so closing. I'm still seeing variability in system CPU usage, so I'm leaving #1675 open, and I'll update that with the latest details.
